File: tests/solutions_js/house_robber_ii.js
class Solution {
    rob(nums) {
        if (!nums.length) return 0;
        if (nums.length === 1) return nums[0];
        if (nums.length === 2) return Math.max(nums[0], nums[1]);

        function robLinear(houses) {
            let prev2 = 0, prev1 = 0;
            for (const num of houses) {
                const curr = Math.max(prev1, prev2 + num);
                prev2 = prev1;
                prev1 = curr;
            }
            return prev1;
        }

        return Math.max(
            robLinear(nums.slice(0, -1)),
            robLinear(nums.slice(1))
        );
    }
}

File: tests/solutions_cpp/house_robber_ii.cpp
#include "json_helper.h"

class Solution {
public:
    int robRange(vector<int>& nums, int lo, int hi) {
        int prev = 0, curr = 0;
        for (int i = lo; i <= hi; i++) {
            int next = max(curr, prev + nums[i]);
            prev = curr;
            curr = next;
        }
        return curr;
    }

    int rob(vector<int>& nums) {
        int n = (int)nums.size();
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        return max(robRange(nums, 0, n - 2), robRange(nums, 1, n - 1));
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    auto nums = to_vector_int(args[0]);
    int result = sol.rob(nums);
    cout << result << endl;
    return 0;
}
